manticore/core/cpu/abstractcpu.py

```diff
@@ -479,89 +571,103 @@ def read_bytes(self, where, size):
 
     #######################################
     # Decoder
-    @abstractmethod
     def _wrap_operands(self, operands):
         '''
         Private method to decorate a capstone Operand to our needs. See Operand
         class
         '''
-        pass
+        raise NotImplementedError
 
     def decode_instruction(self, pc):
         '''
-        This will decode an instruction from memory pointed by `pc` and store
-        it in self.instruction.
+        This will decode an instruction from memory pointed by `pc`
 
         :param int pc: address of the instruction
         '''
-        # No dynamic code!!! #TODO!
-
-        if issymbolic(pc):
-            raise SymbolicPCException()
-
-        if not self.memory.access_ok(pc,'x'):
-            raise InvalidPCException(pc)
-
-        #Check if instruction was already decoded
-        self._instruction_cache = {}
+        #No dynamic code!!! #TODO! 
+        #Check if instruction was already decoded 
         if pc in self._instruction_cache:
             logger.debug("Intruction cache hit at %x", pc)
             return self._instruction_cache[pc]
 
+        def _read_concrete_byte(address):
+            #This reads a byte from memory ignoring permissions
+            #and concretize it if symbolic
+            c = self.memory[address]
+            if issymbolic(c):
+                assert isinstance(c, BitVec) and  c.size == 8
+                if isinstance(c, Constant):
+                    c = chr(c.value)
+                else:
+                    logger.error('Concretize executable memory %r %r', c, text )
+                    raise ConcretizeMemory(self.memory, address = pc,
+                                            size = 8 * self.max_instr_width, 
+                                            policy = 'INSTRUCTION' )
+            assert isinstance(c, str)
+            return c
+
         text = ''
         try:
-            # check access_ok
+            # Read Instruction from memory
             for i in xrange(0, self.max_instr_width):
-                c = self.memory[pc+i]
-                if issymbolic(c):
-                    assert isinstance(c, BitVec) and  c.size == 8
-                    if isinstance(c, Constant):
-                        c = chr(c.value)
-                    else:
-                        logger.error('Concretize executable memory %r %r', c, text )
-                        break
-                assert isinstance(c, str)
-                text += c
-        except MemoryException:
+                text += _read_concrete_byte(pc+i)
+        except MemoryException as mem_except:
+            #lets not raise this exception in case the instruction
+            #was fully read befor the page
```

Comment:
Perhaps rewrite as:
```
try:
    # Read Instruction from memory
    for i in xrange(0, self.max_instr_width):
        try:
            text += _read_concrete_byte(pc+i)
        except MemoryException:
            # In case of reading into unmapped memory
            break
```
Otherwise, we throw+except for every byte until `max_instr_width`.